vernac/comDefinition.ml

```diff
@@ -95,17 +95,26 @@ let interp_definition ~program_mode env evd impl_env bl red_option c ctypopt =
   let evd, c, imps, tyopt =
     match tyopt with
     | None ->
-      let evd, (c, impsbody) = interp_constr_evars_impls ~program_mode ~impls env_bl evd c in
+      let evd, (c, impsbody) =
+        interp_constr_evars_impls_gen ~flags ~impls env_bl evd Pretyping.WithoutTypeConstraint c
+      in
       evd, c, imps1@impsbody, None
     | Some (ty, impsty) ->
-      let evd, (c, impsbody) = interp_casted_constr_evars_impls ~program_mode ~impls env_bl evd c ty in
+      let evd, (c, impsbody) =
+        interp_constr_evars_impls_gen ~flags ~impls env_bl evd (Pretyping.OfType ty) c
+      in
       check_imps ~impsty ~impsbody;
       evd, c, imps1@impsty, Some ty
   in
+  (* solve typeclasses / heuristics *)
+  (* even with program_mode=false we don't fail here as we don't want
+     to fail if there are unused evars *)
+  let solve_flags = Pretyping.{all_no_fail_flags with program_mode} in
+  let evd = Pretyping.solve_remaining_evars solve_flags env evd in
```

Comment:
@herbelin I'm not sure about using interp_definition with this added solve in record (now that #13166 is merged), WDYT?
Maybe separate a interp_definition_core with about this https://github.com/coq/coq/blob/031160d00a8e2cfcd74709a487be260378bad0f2/vernac/comDefinition.ml#L87-L108 ?

Reply:
> @herbelin I'm not sure about using interp_definition with this added solve in record (now that #13166 is merged), WDYT?

Indeed, the solve would come too early.

> Maybe separate a interp_definition_core with about this 

Or `interp_definition_component`? Or keep the `interp_definition` name and add a `interp_standalone_definition`/`interp_complete_definition` with the solve.